### PR TITLE
Feature/1 experimental port to busted testing

### LINKED ISSUE #1: Extract common library from RLSL
- [ ] Create the standard layout of the repository
- [ ] Migrate code and tests from exasol/row-level-security-lua
- [ ] Refactor for use as library
- [ ] Add developer guide

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #1.

### CODE REVIEW

File: .github/workflows/broken_links_checker.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+on:
+  schedule:
+    - cron: "0 5 * * *"
+  push:
+
+jobs:
+  linkChecker:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (kaklakariada) [inline .github/workflows/broken_links_checker.yml:12]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/ci-build.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+on:
+  - push
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout the repository
+        uses: actions/checkout@v2
```
Comment (kaklakariada) [inline .github/workflows/ci-build.yml:11]:
```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/dependencies_check.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+on:
+  schedule:
+    - cron: "0 2 * * *"
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
```
Comment (kaklakariada) [inline .github/workflows/dependencies_check.yml:12]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/dependencies_check.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+name: Dependencies Check
```
Comment (kaklakariada) [inline .github/workflows/dependencies_check.yml:1]: Is this necessary? Unless this also checks lua dependencies it's a wast of resources.

File: dependencies.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+## Runtime Dependencies
+
+| Dependency                                             | License     |
+|--------------------------------------------------------|-------------|
+| [Lua 5.4](https://www.lua.org)                         | [MIT][mit]  |
+| [lua-cjson](https://github.com/openresty/lua-cjson)    | [MIT][mit]  |
+| [remotelog](https://github.com/exasol/remotelog-lua)   | [MIT][mit]  |
+
+`lua-cjson` depends on the `cjson` library both are preinstalled on Exasol.
```
Comment (kaklakariada) [inline dependencies.md:16]:
```suggestion
`lua-cjson` depends on the `cjson` library, both are preinstalled on Exasol.
```

File: doc/developer_guide/developer_guide.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+```bash
+sudo luarocks install LuaUnit
+sudo luarocks install Mockagne
+sudo luarocks install lua-cjson
+sudo luarocks install remotelog
+sudo luarocks install luacov
+sudo luarocks install luacov-coveralls
+sudo luarocks install luacheck
+```
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:41]: Use

```
luarocks install --local --deps-only *.rockspec
```
if possible

File: doc/developer_guide/developer_guide.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+The `luacov` and `luacov-coveralls` libraries take care of measuring and reporting code coverage in the tests.
+
+## How to Run Lua Unit Tests
+
+### Run Unit Tests From Terminal
+
+To run unit tests from terminal, you first need to install Lua:
+
+```bash
+sudo apt install lua5.1
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:54]:
```suggestion
sudo apt install lua5.4
```

File: doc/developer_guide/developer_guide.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+```
+
+Another important thing to do, you need to add the project's directories with lua files to LUA_PATH environment variable.
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:57]:
```suggestion
Another important thing to do, you need to add the project's directories with lua files to `LUA_PATH` environment variable.
```

File: doc/developer_guide/developer_guide.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+```bash
+lua src/test/lua/exasolvs/test_query_renderer.lua 
+```
+
+If you want to run all unit tests including code coverage and static code analysis, issue the following command:
+
+```bash
+tools/runtests.sh
+```
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:74]: Update this to use `busted`

File: doc/developer_guide/developer_guide.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+1. VSCL provides a base library for writing your own Virtual Schemas in Lua
+1. The resulting package is available as LuaRocks package `virtual-schema-common-lua`
+1. In your concrete Virtual Schema implementation you need to write an `entry` module, that has an `adapter_call` entry function
+1. The `entry` module should create and wire up all static objects (like the `RequestDispatcher` for example)
+1. Use the `remotelog` package for logging
+
+### Running the Unit Tests From Intellij IDEA
+
+First, you need to install a plug-in that handles Lua code. We recommend to use `lua` plugin by `sylvanaar`.
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:100]: I thought it was EmmyLua?

File: doc/developer_guide/developer_guide.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+In the next step we add a Lua interpreter. Fow that go to `File` &rarr; `Project structure` &rarr; `Modules`.
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:102]:
```suggestion
In the next step we add a Lua interpreter. For that go to `File` &rarr; `Project structure` &rarr; `Modules`.
```

File: doc/developer_guide/developer_guide.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+Here press `Add` button in the upper left corner and add a new Lua framework.
+You can use one of the default Lua interpreters suggested by Intellij or add your own in `SDKs` tab of the `Project structure`.
+We recommend installing and using `lua5.1`.
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:105]:
```suggestion
We recommend installing and using `lua5.4`.
```

File: spec/QueryRenderer_spec.lua [hunk 13]
```diff
@@ -0,0 +1,1567 @@
...
+describe("QueryRenderer", function()
+    it("renders SELECT *", function()
+        local original_query = {
+            type = "select",
+            from = {type  = "table", name = "T1"}
+        }
+        assert.renders_to('SELECT * FROM "T1"', original_query )
+    end)
+
+    it("renders a SELECT with a table an two columns", function()
```
Comment (kaklakariada) [inline spec/QueryRenderer_spec.lua:14]:
```suggestion
    it("renders a SELECT with a table and two columns", function()
```

File: spec/QueryRenderer_spec.lua [hunk 14]
```diff
@@ -0,0 +1,1567 @@
...
+require("spec.assertions.renderer_assertions")
+local QueryRenderer = require("exasolvs.QueryRenderer")
+
+describe("QueryRenderer", function()
+    it("renders SELECT *", function()
+        local original_query = {
+            type = "select",
+            from = {type  = "table", name = "T1"}
+        }
+        assert.renders_to('SELECT * FROM "T1"', original_query )
```
Comment (kaklakariada) [inline spec/QueryRenderer_spec.lua:11]:
```suggestion
        assert.renders_to('SELECT * FROM "T1"', original_query)
```

File: spec/QueryRenderer_spec.lua [hunk 15]
```diff
@@ -0,0 +1,1567 @@
...
+                                   type = "predicate_greater",
+                                   left = {type = "column", name="DAYS_IN_MONTH", tableName = "MONTHS"},
+                                   right = {type = "literal_exactnumeric", value = "30"}
+                               }
+                }
+            }
+        }
+        assert.renders_to('SELECT "MONTHS"."NAME" FROM "MONTHS" WHERE ((\'Q3\' = "MONTHS"."QUARTER") '
+                .. 'AND ("MONTHS"."DAYS_IN_MONTH" > 30))', original_query)
+
```
Comment (kaklakariada) [inline spec/QueryRenderer_spec.lua:107]:
```suggestion
```

File: spec/QueryRenderer_spec.lua [hunk 16]
```diff
@@ -0,0 +1,1567 @@
...
+                            }
+                        }
+                    }
+                }
+                assert.renders_to(parameter.expected, original_query)
+            end)
+        end
+    end)
+
+    describe("renders a scalar function in the SELECT list with a three arguments:", function()
```
Comment (kaklakariada) [inline spec/QueryRenderer_spec.lua:1054]:
```suggestion
    describe("renders a scalar function in the SELECT list with three arguments:", function()
```

File: spec/QueryRenderer_spec.lua [hunk 17]
```diff
@@ -0,0 +1,1567 @@
...
+                            arguments = { { type = parameter.arg_type, value = parameter.arg_value } }
+                        }
+                    }
+                }
+                assert.renders_to(parameter.expected, original_query)
+            end)
+        end
+    end)
+
+    describe("renders a scalar function in the SELECT list with a two arguments:", function()
```
Comment (kaklakariada) [inline spec/QueryRenderer_spec.lua:590]:
```suggestion
    describe("renders a scalar function in the SELECT list with two arguments:", function()
```

File: spec/QueryRenderer_spec.lua [hunk 18]
```diff
@@ -0,0 +1,1567 @@
...
+                                    }
+                                }
+                            }
+                        }
+                        assert.renders_to(parameter.expected, original_query)
+                    end)
+        end
+    end)
+
+    describe("renders a scalar function in the SELECT list with a four arguments:", function()
```
Comment (kaklakariada) [inline spec/QueryRenderer_spec.lua:1111]:
```suggestion
    describe("renders a scalar function in the SELECT list with four arguments:", function()
```

File: spec/QueryRenderer_spec.lua [hunk 19]
```diff
@@ -0,0 +1,1567 @@
...
+                            }
+                        }
+                    }
+                }
+                assert.renders_to(parameter.expected, original_query)
+            end)
+        end
+    end)
+
+    describe("renders a the scalar function EXTRACT", function()
```
Comment (kaklakariada) [inline spec/QueryRenderer_spec.lua:1152]:
```suggestion
    describe("renders the scalar function EXTRACT", function()
```

File: spec/assertions/renderer_assertions.lua [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+local QueryRenderer = require("exasolvs.QueryRenderer")
+
+local function renders_to(_, arguments)
+    local expected = arguments[1]
+    local original_query = arguments[2]
+    local renderer = QueryRenderer.create(original_query)
+    local ok, result = pcall(renderer.render, renderer)
+    arguments[1] = original_query
+    arguments[2] =expected
```
Comment (kaklakariada) [inline spec/assertions/renderer_assertions.lua:14]:
```suggestion
    arguments[2] = expected
```

File: spec/text_spec.lua [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+        it("ignores padding with spaces", function()
+            assert.are.same({"Hello", "world", "!"}, text.split(" Hello , world    , ! "))
+        end)
+
+        it("returns empty array given empty string", function()
+            assert.are.same({}, text.split(""))
+        end)
+
+        it("returns empty nil given nil", function()
```
Comment (kaklakariada) [inline spec/text_spec.lua:44]:
```suggestion
        it("returns nil given nil", function()
```

File: tools/runtests.sh [hunk 22]
```diff
@@ -0,0 +1,118 @@
+#!/bin/bash
```
Comment (kaklakariada) [inline tools/runtests.sh:1]: This file should not be necessary with `busted`?

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: doc/developer_guide/developer_guide.md (hunk 9 above)
Comment (redcatbear) [reply doc/developer_guide/developer_guide.md:74]: Done.

File: doc/developer_guide/developer_guide.md (hunk 10 above)
Comment (redcatbear) [reply doc/developer_guide/developer_guide.md:100]: It is. I will change that in another ticket: #5

File: tools/runtests.sh (hunk 22 above)
Comment (redcatbear) [reply tools/runtests.sh:1]: Removed.

File: .github/workflows/dependencies_check.yml (hunk 4 above)
Comment (redcatbear) [reply .github/workflows/dependencies_check.yml:1]: Removed.

File: doc/developer_guide/developer_guide.md (hunk 6 above)
Comment (redcatbear) [reply doc/developer_guide/developer_guide.md:41]: Done. I might have to distinguish between developer tools and real dependencies later (#5).

File: doc/developer_guide/developer_guide.md [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+```bash
+luarocks path >> ~/.bashrc
+```
+
+### Installing the Required Lua Packages
+
+You need the packages for unit testing, mocking and JSON processing.
+
+Execute as `root` or modify to install in your home directory:
```
Comment (kaklakariada) [inline doc/developer_guide/developer_guide.md:31]:
```suggestion
Execute this to install in your home directory:
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**